crates/router/src/core/payment_methods.rs

```
@@ -2650,3 +2683,196 @@ impl pm_types::SavedPMLPaymentsInfo {
+                                            );
+
+    let key = payment_method_data::SingleUseTokenKey::store_key(&payment_method.id);
+
+    add_single_use_token_to_store(&state, key, value)
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to store single use token")?;
+
+    Ok(())
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+async fn add_single_use_token_to_store(
+    state: &SessionState,
+    key: payment_method_data::SingleUseTokenKey,
+    value: payment_method_data::SingleUsePaymentMethodToken,
+) -> CustomResult<(), errors::StorageError> {
+    let redis_connection = state
+        .store
+        .get_redis_conn()
+        .map_err(Into::<errors::StorageError>::into)?;
+
+    redis_connection
+        .serialize_and_set_key_with_expiry(
+            &payment_method_data::SingleUseTokenKey::get_store_key(&key).into(),
+            value,
+            consts::DEFAULT_PAYMENT_METHOD_STORE_TTL,
+        )
+        .await
+        .change_context(errors::StorageError::KVError)
+        .attach_printable("Failed to insert payment method token to redis")?;
+    Ok(())
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+async fn get_single_use_token_from_store(
+    state: &SessionState,
+    key: payment_method_data::SingleUseTokenKey,
+) -> CustomResult<Option<payment_method_data::SingleUsePaymentMethodToken>, errors::StorageError> {
+    let redis_connection = state
+        .store
+        .get_redis_conn()
+        .map_err(Into::<errors::StorageError>::into)?;
+
+    match redis_connection
+        .get_and_deserialize_key::<payment_method_data::SingleUsePaymentMethodToken>(
+            &payment_method_data::SingleUseTokenKey::get_store_key(&key).into(),
+            "SingleUsePaymentMethodToken",
+        )
+        .await
+    {
+        Ok(token) => Ok(Some(token)),
+        Err(e) => Ok(None),
```

do not ignore this error.